hoomd/hpmc/pytest/test_remove_drift.py

```diff
@@ -0,0 +1,189 @@
+
+
+@pytest.mark.cpu
+@pytest.mark.parametrize("attr,value", valid_attrs)
+def test_valid_setattr_attached(attr, value, simulation_factory,
+                                two_particle_snapshot_factory, valid_args):
+    """Test that RemoveDrift can get and set attributes while attached."""
+    integrator = valid_args[0]
+    args = valid_args[1]
+    # Need to unpack union integrators
+    if isinstance(integrator, tuple):
+        inner_integrator = integrator[0]
+        integrator = integrator[1]
+        inner_mc = inner_integrator()
+        for i in range(len(args["shapes"])):
+            # This will fill in default values for the inner shape objects
+            inner_mc.shape["A"] = args["shapes"][i]
+            args["shapes"][i] = inner_mc.shape["A"]
+    mc = integrator()
+    mc.shape["A"] = args
+    mc.shape["B"] = args
+
+    remove_drift = hoomd.hpmc.update.RemoveDrift(
+        trigger=hoomd.trigger.Periodic(10),
+        reference_positions=[(0, 0, 1), (-1, 0, 1)])
+    dim = 2 if 'polygon' in integrator.__name__.lower() else 3
+    sim = simulation_factory(
+        two_particle_snapshot_factory(particle_types=['A', 'B'],
+                                      dimensions=dim,
+                                      d=2,
+                                      L=50))
+    sim.operations.updaters.append(remove_drift)
+    sim.operations.integrator = mc
+
+    sim.run(0)
+
+    setattr(remove_drift, attr, value)
+    assert np.all(getattr(remove_drift, attr) == value)
+
+
+@skip_mpi
+@pytest.mark.cpu
+def test_remove_drift(simulation_factory, lattice_snapshot_factory):
+    """Test that RemoveDrift modifies positions correctly."""
+    sim = simulation_factory(
+        lattice_snapshot_factory(particle_types=['A'],
+                                 dimensions=3,
+                                 a=4,
+                                 n=10,
+                                 r=0))
+    sim.seed = 19233
+    mc = hoomd.hpmc.integrate.Sphere(default_d=0.1, default_a=0.1)
+    mc.shape["A"] = dict(diameter=1.0)
+    sim.operations.integrator = mc
+
+    # use initial lattice configuration as reference
+    comm = MPI.COMM_WORLD
+    s = sim.state.snapshot
+    if s.communicator.rank == 0:
+        reference_positions = s.particles.position
+    else:
+        reference_positions = None
+    reference_positions = comm.bcast(reference_positions, root=0)
+
+    # randomize a bit
+    sim.run(100)
+
+    # remove the drift from the previous run
+    remove_drift = hoomd.hpmc.update.RemoveDrift(
+        trigger=hoomd.trigger.Periodic(1),
+        reference_positions=reference_positions)
+    sim.operations.updaters.append(remove_drift)
+    sim.run(1)
+
+    s = sim.state.snapshot
+    if s.communicator.rank == 0:
+        reference_com = np.mean(reference_positions, axis=0)
+        new_com = np.mean(s.particles.position, axis=0)
+        assert np.allclose(reference_com, new_com, atol=0.05)
```

Comment:
This test is failing. Given the periodic boundary conditions, I don't think that the remove drift updater leads to a conservation of the center of mass of the system.

To test, you need to compute the drift (you can avoid periodic boundary conditions by unwrapping positions) from the initial lattice position and ensure that the drift is close to 0.

Reply:
I see. I thought it was related to the particle tags.

Reply:
@joaander I changed the test to compute the drift from the original configuration (after unwrapping) and check if its close to zero but it keeps failing. I've checked the source code for any obvious issues but nothing stands out to me.

Reply:
To understand why this test is failing, I would suggest inspecting the before and after remove drift configurations in a visualization tool.